### PR TITLE
feat(upload): upload component

### LINKED ISSUE #128: ### Upload
Assignee @JeremyWuuuuu

### PR DESCRIPTION
**Pull request by jw-foss**
- [x] Component Migration

Fix #128 

Still WIP
Tests were not added yet

### CODE REVIEW

File: packages/test-utils/make-mount.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { mount } from '@vue/test-utils'
+import { merge } from 'lodash'
+
+const makeMount = <C,O, E>(element: C, defaultOptions: O) => {
```
Comment (JeremyWuuuuu) [inline packages/test-utils/make-mount.ts:4]: Shoulda add a space after the comma.

File: packages/upload/index.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { App } from 'vue'
+import Upload from './src/index.vue'
+export default (app: App): void => {
+  app.component(Upload.name, Upload)
+}
```
Comment (g3r4n) [inline packages/upload/index.ts:5]:
```suggestion
}
export const ElUpload  = Upload;
```

File: packages/upload/src/ajax.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
+import type {
+  ElUploadProgressEvent,
+  ElUploadRequestOptions,
+  ElUploadAjaxError,
+} from './upload'
+
+function getError(action: string, option: ElUploadRequestOptions, xhr: XMLHttpRequest) {
```
Comment (g3r4n) [inline packages/upload/src/ajax.ts:7]:
```suggestion
function getError(action: string, option: ElUploadRequestOptions, xhr: XMLHttpRequest) : string | ElUploadAjaxError {
```
Should it be like that ? that's the first time I'm doing TypeScript 😛

File: packages/upload/doc/basic.vue [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    :file-list="fileList"
+  >
+    <el-button size="small" type="primary">Click to upload</el-button>
+    <template #tip>
+      <div class="el-upload__tip">jpg/png files with a size less than 500kb</div>
+    </template>
+  </el-upload>
+</template>
+<script lang="ts">
+export default {
```
Comment (g3r4n) [inline packages/upload/doc/basic.vue:20]: Should we use a setup function or <script lang="ts" setup> ?

File: packages/upload/index.ts (hunk 2 above)
Comment (g3r4n) [inline packages/upload/index.ts:5]:
```suggestion
}
export const ElUpload = Upload;
```

File: packages/upload/src/ajax.ts (hunk 3 above)
Comment (g3r4n) [inline packages/upload/src/ajax.ts:7]:
```suggestion
function getError(action: string, option: ElUploadRequestOptions, xhr: XMLHttpRequest) : ElUploadAjaxError {
```
Is it ok do that ? Learning TypeScript as I review the PR.

File: packages/upload/src/upload.vue [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+      type="file"
+      :name="name"
+      :multiple="multiple"
+      :accept="accept"
+      @change="handleChange"
+    >
+  </div>
+</template>
+
+<script lang="tsx">
```
Comment (g3r4n) [inline packages/upload/src/upload.vue:28]:
```suggestion
<script lang="ts">
```
why tsx ?

File: packages/test-utils/make-mount.ts (hunk 1 above)
Comment (HerringtonDarkholme) [reply packages/test-utils/make-mount.ts:4]: I don't think the function covers most use case in the code base, though.

Comment (JeremyWuuuuu) [reply packages/test-utils/make-mount.ts:4]: Agreed, we have no idea what situation in the future we might encounter, let's keep this function evolving.

File: packages/upload/src/ajax.ts (hunk 3 above)
Comment (HerringtonDarkholme) [reply packages/upload/src/ajax.ts:7]: No need to explicit annotate return type for internal function

File: packages/upload/__tests__/upload.spec.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+const AXIOM = 'Rem is the best girl'
+const action = 'test-action'
+
+interface MockFile {
+  name: string
+  body: string
+  mimeType: string
+}
+
+const mockGetFile = (element: HTMLInputElement, files: File[]) => {
```
Comment (HerringtonDarkholme) [inline packages/upload/__tests__/upload.spec.ts:17]: How about patch HTMLInputElement?

File: packages/upload/src/ajax.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  let msg: string
+  if (xhr.response) {
+    msg = `${xhr.response.error || xhr.response}`
```
Comment (HerringtonDarkholme) [inline packages/upload/src/ajax.ts:10]:
```suggestion
    msg = String(xhr.response.error || xhr.response)
```

File: packages/upload/src/index.vue [hunk 8]
```diff
@@ -0,0 +1,247 @@
+<script lang='ts'>
```
Comment (HerringtonDarkholme) [inline packages/upload/src/index.vue:1]: Why not just use .ts file?

File: packages/upload/src/index.vue [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+  FileHandler,
+  FileResultHandler,
+} from './upload'
+
+type PFileHandler<T> = PropType<FileHandler<T>>
+type PFileResultHandler<T = any> = PropType<FileResultHandler<T>>
+
+export default defineComponent({
+  name: 'ElUpload',
+  components: {
```
Comment (HerringtonDarkholme) [inline packages/upload/src/index.vue:32]: Is this registry used?

File: packages/upload/src/index.vue [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+      'on-success': this.handleSuccess,
+      'on-error': this.handleError,
+      'on-preview': this.onPreview,
+      'on-remove': this.handleRemove,
+      'http-request': this.httpRequest,
+      ref: 'uploadRef',
+    }
+    const trigger = this.$slots.trigger || this.$slots.default
+    const uploadComponent = h(Upload, uploadData, {
+      default: () => trigger?.(),
```
Comment (HerringtonDarkholme) [inline packages/upload/src/index.vue:235]: Consider point-free style

File: packages/upload/src/index.vue [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+      handleStart,
+      handleSuccess,
+      handleRemove,
+      submit,
+      uploadRef,
+      uploadFiles,
+    } = useHandlers(props)
+
+
+    provide('uploader', getCurrentInstance())
```
Comment (HerringtonDarkholme) [inline packages/upload/src/index.vue:155]: .proxy

**Review by HerringtonDarkholme — APPROVED** [review]

File: packages/upload/doc/basic.vue (hunk 4 above)
Comment (HerringtonDarkholme) [reply packages/upload/doc/basic.vue:20]: Nope, setup script is still experimental and cannot properly handle props.

**Pull request merged**